src/torrent.cpp

```diff
@@ -4070,7 +4070,8 @@ namespace {
 					we_have(verified_piece);
 				}
 			}
-			else if (result.status == set_block_hash_result::result::block_hash_failed)
+			else if (result.status == set_block_hash_result::result::block_hash_failed
+				|| result.status == set_block_hash_result::result::piece_hash_failed)
```

Comment:
This is in a loop which sets each block hash individually so there may be a spurious `piece_hash_failed` in the middle of the loop if any of the yet-to-be-set block hashes are non-zero (e.g. after re-downloading a previously failed piece). Only a piece hash failure on the last block can be considered a definite failure.

Reply:
makes sense. given the `block_hashes` are guaranteed to cover exactly one piece, is it ever possible for the loop, in the case `result.status == set_block_hash_result::result::success`, to take more than one turn? and is it ever possible for `verified_piece != piece`?